\documentclass{article}

\usepackage{fontspec}
\defaultfontfeatures{Scale=MatchLowercase,Mapping=tex-text}
\setmainfont[Ligatures=TeX]{DejaVu Serif}
\setsansfont[Ligatures=TeX]{DejaVu Sans}
\setmonofont{DejaVu Sans Mono}

\linespread{1.05} % Line spacing - Palatino needs more space between lines
\usepackage{microtype} % Slightly tweak font spacing for aesthetics
\usepackage{hyperref} % For hyperlinks in the PDF
\usepackage{adjustbox} % For resizing
\usepackage{graphicx} % For spaces in pictures
\usepackage[space]{grffile} % Same
\usepackage[margin=0.5in]{geometry} % Adjust margins
\usepackage[english]{babel} % Specify a different language here - english by default
\usepackage{adjustbox} % For resizing
% 
\usepackage{listings}
% Proof package
\usepackage{amsthm}
\usepackage{tikz}
%----------------------------------------------------------------------------------------
%	TITLE SECTION
%----------------------------------------------------------------------------------------

\title{\vspace{-15mm}\fontsize{24pt}{10pt}\selectfont\textbf{Ultra-low input ChIP-Seq processing pipeline}} % Article title

\author{
\large
\textsc{Oleg Shpynov}\\
\normalsize \href{mailto:oleg.shpynov@jetbrains.com}{oleg.shpynov@jetbrains.com} % Your email address
\vspace{-1mm}
}
\date{\today}

\begin{document}

\maketitle % Print the title and abstract box

%----------------------------------------------------------------------------------------
%	ABSTRACT
%----------------------------------------------------------------------------------------
\abstract{
Epigenetics plays a crucial role in processes of differentiation, aging, deceases, etc. Different epigenetics modifications and their combinations mediate biological processes such as transcription cooperatively. In this project we are interested in different epigenetic marks during healthy human aging process. We collect and profile histone modifications in blood cell monocytes in 2 cohorts of people Young and Old, 20 persons in each.
From computational point of view we are going to align the raw reads to human genome build hg19, process Quality Control analysis: reads quality, length, alignment rate, PCR artifacts, strand correlation, alignment to blacklisted regions provided by ENCODE, etc. Peak calling and differential peak calling is also on the list. We are particularly interested in ChIP-Seq peaks distribution: how many of peaks we observe in each condition, what are the positions, which are donor specific; how do they compare among different histone modification and open chromatin data. 
We plan to try different tools and and approaches: \textit{FastQC} for reads quality, alignment with \textit{Bowtie2} and \textit{BWA}, different peak calling tools for broad and narrow peaks - \textit{SICER} and \textit{MACS2}, differential peak calling algorithms like \textit{ChipDiff}, \textit{MACS2bgdiff}, \textit{MAnorm}, \textit{ZINBRA} etc. \\
Here we present effective paralleled pipeline for Ultra Low ChIP-Seq data processing.
}

%----------------------------------------------------------------------------------------
%	ARTICLE CONTENTS
%----------------------------------------------------------------------------------------

\tableofcontents % Print the contents section¬
\thispagestyle{empty} % Removes page numbering from the first page
\clearpage
\setcounter{page}{1}

\section{Ultra Low Input ChIP-Seq}
Chromatin immunoprecipitation followed by high-throughput sequencing (ChIP-seq) is a technique that provides quantitative, genome-wide mapping of target protein binding events. Identifying putative protein binding sites from large, sequence based datasets presents a bioinformatic challenge that has required considerable computational innovation despite the availability of numerous programs for ChIP-Chip analysis. Conventional histone mark ChIP-Seq experiment requires at least 1 million cells, while Ultra-Low input ChiP-Seq (ULI-ChIP) can perform well with 100 thousand cells.

\section{Data}
We used ULI-ChIP experimental protocol for sequencing of 40 samples (20 Older and 20 Younger) in 2 batches. 
Sample ID consists of 2 parts – Age (denoted with \textit{YD}/\textit{OD}) and number. Samples with numbers $\leq$ 10 belong to the 1st batch.\\
All the ChIP-Seq data is located within \texttt{/scratch/artyomov\_lab\_aging/Y10OD10/chipseq/}
\begin{lstlisting}
> tree -L 1 /scratch/artyomov_lab_aging/Y10OD10/chipseq/
/scratch/artyomov_lab_aging/Y10OD10/chipseq/
|-- indexes
|-- processed
|-- raw-reads 
\end{lstlisting}
Folder \texttt{indexes} contains all the necessary indexes for ChIP-Seq pipeline, \texttt{processed} contains processing results and \texttt{raw-reads} is the most important one with raw data in FASTQ format.

\section{Raw-data}
Sickkids sequencing facility provides data in FASTQ format and it is accessible for download using sftp protocol. Data package contains untrimmed reads in FASTQ format, generated using \textit{bcl2fastq2} v2.17.\\
As an example we show file structure for K27ac raw data after downloading, md5 checking and unzipping:
\begin{lstlisting}
> tree /scratch/artyomov_lab_aging/Y10OD10/chipseq/raw-reads/k27ac_10vs10_LOG1389/
k27ac_10vs10_LOG1389/
|-- 170223_D00430_0239_BCA8JLANXX
|	|-- md5_real.txt
|	|-- md5.txt
|	|-- OD10_R1.fastq
|	|-- OD10_R1.fastq.gz.md5
|	|-- ...
|-- readme_disclaimer.txt
\end{lstlisting}

\section{Pipeline requirements}


\section{Technical ULI-ChIP pipeline}
All the scripts necessary for ULI-ChIP processing are available on github in the dedicated private repository\url{https://github.com/JetBrains-Research/washu}.\\ 
Technical processing consists of the following steps:
\begin{itemize}
\item QC
\item Trimming
\item Alignment
\item Coverage visualization
\item Optional subsampling
\item Peak calling
\end{itemize}
QC step uses \textit{fastqc} and \textit{mutliqc} for quality control and aggregated statistics. Alignment step uses \textit{bowtie}, \textit{bowtie2} for ChIP-Seq reads or \textit{star} for RNA-Seq reads alignment. Alignment statistics is collected automatically and is presented in a log file \texttt{bowtie\_report.csv}. Coverage visualization step builds reads coverage using \textit{bedtools} and \textit{bedGraphToBigWig}, both as is and deduplicated tracks are available as results. Peak calling step uses \textit{MACS2}, \textit{MACS14} or \textit{RSEG} peak callers with various settings and produce peaks, and signal tracks when input ChIP-Seq is available. File \texttt{macs2\_report.csv} contains summary statistics of peak calling including Fraction of Reads In Reads for each sample.\\ 

We use processed and raw-data separation to prevent possible raw-data corruption, so generally raw-data folders are marked as read-only.
\begin{lstlisting}
> ls -lah /scratch/artyomov_lab_aging/Y10OD10/chipseq/raw-reads/
dr-xr-xr-x 3 oshpynov martyomov 4.0K Mar  4 00:31 k27ac_10vs10_LOG1389
dr-xr-xr-x 3 oshpynov martyomov 4.0K Feb 14 08:12 k4me1_10vs10_LOG1287
\end{lstlisting}

As a preliminary step we use soft links (\texttt{ln -s source target}) to prepare data. E.g. all the FASTQ files in \texttt{/scratch/artyomov\_lab\_aging/Y10OD10/chipseq/processed/k27ac\_10vs10} are soft references to \texttt{/scratch/artyomov\_lab\_aging/Y10OD10/chipseq/raw-reads/k27ac\_10vs10\_LOG1389}.\\

Command line for technical pipeline:
\begin{lstlisting}
> python3 pipeline_chipseq.py /scratch/artyomov_lab_aging/Y10OD10/chipseq/processed/k27ac_10vs10 
\end{lstlisting}

Pipeline uses \textit{convention over configuration} approach, i.e. it encodes result folders using self-describing suffixes, i.e. \texttt{\_bam} for alignment folder. The following folders are created:
\begin{lstlisting}
> ls /scratch/artyomov_lab_aging/Y10OD10/chipseq/processed | grep k27ac_10vs10
k27ac_10vs10
k27ac_10vs10_bams
k27ac_10vs10_bams_bws
k27ac_10vs10_bams_macs_0.01
k27ac_10vs10_bams_macs_broad_0.01
\end{lstlisting}

All the necessary indexes are downloaded and built automatically upon request. All the steps are highly paralleled so that typical technical step can processed in a couple of hours depending on computational cluster load, but pipeline can be launched locally as well.

\section{Differential ChIP-Seq analysis}
Differential analysis used Diffbind configuration to process \textit{MACS2}, \textit{ChIPDiff} and \textit{Diffbind} analysis from a single script.
See \texttt{analysis/diffbind\_config.sh} to create Diffbind configuration using naming conventions.\\
Command line to create config for K27ac in \texttt{k27ac\_10vs10\_diff}:\\
\begin{lstlisting}
> bash analysis/diffbind_config.sh /scratch/artyomov_lab_aging/Y10OD10/chipseq/processed k27ac_10vs10 0.01\
  k27ac_10vs10_diff/diffbind.csv
\end{lstlisting}

And finally script to launch analysis:
\begin{lstlisting}
> bash analysis/chipseq_diff.sh diff_27ac_10vs10 /scratch/artyomov_lab_aging/Y10OD10/chipseq/indexes/hg19/hg19.chrom.sizes\
 /scratch/artyomov_lab_aging/Y10OD10/chipseq/indexes/hg19/Homo_sapiens.GRCh37.87.gtf.gz\
 k27ac_10vs10_diff/diffbind.csv
\end{lstlisting}
\section{Results synchronization}
\texttt{RSync} is a utility for efficiently transferring and synchronizing files across computer systems. We use \texttt{rsync} over ssh to download technical pipeline results locally for further analysis.\\
The following command can be used to download all the \texttt{csv} files locally:
\begin{lstlisting}
> rsync -aivz --progress -e "ssh -c arcfour -i <PRIVATE_SSH_KEY> -o StrictHostKeyChecking=no\
 -o UserKnownHostsFile=/dev/null" <USERNAME>@dtn01.chpc.wustl.edu:/scratch/artyomov_lab_aging/Y10OD10/chipseq/\
 <LOCAL_PATH> --include=*.csv --exclude=*.* 
\end{lstlisting}

\section{Useful scripts}
\begin{itemize}
\item \texttt{abf/generate\_browser.py} Generates static html files for simple results browser
\item \texttt{bed/bedtrace.py} \textit{pybedtools} like python library to process multiple BED comparison
\item \texttt{bed/chromhmm\_stats.sh} Computes overlap statistics with \textit{ChromHMM} markup
\item \texttt{bed/closest\_gene.sh} Computes closest gene for each position in given BED file
\item \texttt{bed/compare.sh} Compares 2 BED files, separates locations in common and different
\item \texttt{bed/intersect.sh} Intersects list of BED files, order is not important in contrary with \textit{pybedtools}
\item \texttt{bed/metapeaks.sh} Compares list of BED files, can be used for Venn diagrams
\item \texttt{bed/minus.sh} Substract 2 BED files
\item \texttt{bed/union.sh} Union list of BED files
\item \texttt{notebooks} Jupyter notebooks with ULI-ChIP data processing
\item \texttt{ucsc} Folder with existing UCSC sessions e.g.: \texttt{pilot.txt} for \url{https://genome.ucsc.edu/cgi-bin/hgTracks?hgS_doOtherUser=submit&hgS_otherUserName=olegshpynov&hgS_otherUserSessionName=pilot}
\item \texttt{tex} Folder with this document sources 
\end{itemize}

\section{Github repo}
See \texttt{README.md} file for information on repository structure.\\
\includegraphics[width=\linewidth]{readme.png}


\end{document}